cmd/kubelet/app/server.go

```diff
@@ -377,116 +347,63 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
-			if err == nil {
-				// Update s (KubeletServer) with new config from API server
-				s.KubeletConfiguration = *remoteKC
-				// Ensure that /configz is up to date with the new config
-				if cfgzErr != nil {
-					glog.Errorf("was unable to register configz before due to %s, will not be able to set now", cfgzErr)
-				} else {
-					setConfigz(cfgz, &s.KubeletConfiguration)
-				}
-				// Update feature gates from the new config
-				err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-				if err != nil {
-					return err
-				}
-			} else {
-				glog.Errorf("failed to init dynamic Kubelet configuration sync: %v", err)
-			}
-		}
-	}
-
-	// Validate configuration.
-	if err := validateConfig(s); err != nil {
-		return err
-	}
-
+	var nodeName types.NodeName
 	if kubeDeps == nil {
-		var kubeClient clientset.Interface
-		var eventClient v1core.EventsGetter
-		var externalKubeClient clientgoclientset.Interface
-		var cloud cloudprovider.Interface
-
-		if !cloudprovider.IsExternal(s.CloudProvider) && s.CloudProvider != componentconfigv1alpha1.AutoDetectCloudProvider {
-			cloud, err = cloudprovider.InitCloudProvider(s.CloudProvider, s.CloudConfigFile)
-			if err != nil {
-				return err
-			}
-			if cloud == nil {
-				glog.V(2).Infof("No cloud provider specified: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
-			} else {
-				glog.V(2).Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
-			}
+		cloud, err := getCloud(s)
+		if err != nil {
+			return err
 		}
-
-		nodeName, err := getNodeName(cloud, nodeutil.GetHostname(s.HostnameOverride))
+		nodeName, err = getNodeName(cloud, nodeutil.GetHostname(s.HostnameOverride))
 		if err != nil {
 			return err
 		}
-
 		if s.BootstrapKubeconfig != "" {
 			if err := bootstrapClientCert(s.KubeConfig.Value(), s.BootstrapKubeconfig, s.CertDirectory, nodeName); err != nil {
 				return err
 			}
 		}
-
-		clientConfig, err := CreateAPIServerClientConfig(s)
-		if err == nil {
-			kubeClient, err = clientset.NewForConfig(clientConfig)
-			if err != nil {
-				glog.Warningf("New kubeClient from clientConfig error: %v", err)
-			}
-			externalKubeClient, err = clientgoclientset.NewForConfig(clientConfig)
-			if err != nil {
-				glog.Warningf("New kubeClient from clientConfig error: %v", err)
-			}
-			// make a separate client for events
-			eventClientConfig := *clientConfig
-			eventClientConfig.QPS = float32(s.EventRecordQPS)
-			eventClientConfig.Burst = int(s.EventBurst)
-			eventClient, err = clientgoclientset.NewForConfig(&eventClientConfig)
-			if err != nil {
-				glog.Warningf("Failed to create API Server client: %v", err)
-			}
-		} else {
-			if s.RequireKubeConfig {
-				return fmt.Errorf("invalid kubeconfig: %v", err)
-			}
-			if standaloneMode {
-				glog.Warningf("No API client: %v", err)
-			}
+		kubeClient, _, _, err := getClients(s, standaloneMode)
```

Comment:
> Are these issues (which it sounds like @mtaufen agrees with) worth delaying for 3 more months?

I think these issues are severe enough to warrant delaying the feature until it is ready.  Tying bits together as I've seen in just this one file (haven't even gotten to the rest of the code) is a step backwards in terms of clarity and makes it even hard to unwind later.



> Are you developing something right now that this makes significantly more difficult to implement? If so, is it something we can work out next quarter?

I don't think that making the code harder to understand and reason about now is made more palatable by claiming, "you aren't touching it right now".  Merging this pull would make it much harder for someone to make changes if they had a reason to.  All code requires maintenance and this would make that burden harder.  

This is an important feature, let's take the time to do it right instead taking short-term gain action now.